pkg/scheduler/queue/tenant_queues.go

```diff
@@ -23,32 +24,66 @@ type tenantRequest struct {
 // queueBroker encapsulates access to tenant queues for pending requests
 // and maintains consistency with the tenant-querier assignments
 type queueBroker struct {
-	tenantQueuesTree *TreeQueue
+	tree *TreeQueue
 
-	tenantQuerierAssignments tenantQuerierAssignments
+	tenantQuerierAssignments *tenantQuerierAssignments
 
 	maxTenantQueueSize               int
 	additionalQueueDimensionsEnabled bool
+	prioritizeQueryComponents        bool
 }
 
-func newQueueBroker(maxTenantQueueSize int, additionalQueueDimensionsEnabled bool, forgetDelay time.Duration) *queueBroker {
-	return &queueBroker{
-		tenantQueuesTree: NewTreeQueue("root"),
-		tenantQuerierAssignments: tenantQuerierAssignments{
-			queriersByID:       map[QuerierID]*querierConn{},
-			querierIDsSorted:   nil,
-			querierForgetDelay: forgetDelay,
-			tenantIDOrder:      nil,
-			tenantsByID:        map[TenantID]*queueTenant{},
-			tenantQuerierIDs:   map[TenantID]map[QuerierID]struct{}{},
-		},
+func newQueueBroker(
+	maxTenantQueueSize int,
+	additionalQueueDimensionsEnabled bool,
+	prioritizeQueryComponents bool,
+	forgetDelay time.Duration,
+) *queueBroker {
+	currentQuerier := QuerierID("")
+	tqas := &tenantQuerierAssignments{
+		queriersByID:       map[QuerierID]*querierConn{},
+		querierIDsSorted:   nil,
+		querierForgetDelay: forgetDelay,
+		tenantIDOrder:      nil,
+		tenantsByID:        map[TenantID]*queueTenant{},
+		tenantQuerierIDs:   map[TenantID]map[QuerierID]struct{}{},
+		currentQuerier:     &currentQuerier,
+		tenantOrderIndex:   localQueueIndex - 1,
+	}
+
+	// If prioritizeQueryComponents is true, the tree will be created
+	// with query components at one level above tenants; if it is false,
+	// tenant nodes will each maintain their own query component subtree.
+	tree, err := NewTree(
+		tqas,               // root
+		&roundRobinState{}, // tenant queues
+		&roundRobinState{}, // query components
+	)
+	if prioritizeQueryComponents {
+		tree, err = NewTree(
+			&roundRobinState{}, // root
+			tqas,               // query components
+			&roundRobinState{}, // tenant queues
+		)
```

Comment:
I find these comments confusing because it's annotating what is in the queues at each level, but the args we are providing are for selecting the the child node, or for making the "decision" at that node if we think of it as a decision tree.

I think of it more like:
```golang
tree, err := NewTree(
		tqas,               // select tenant
		&roundRobinState{}, // select query component
		&roundRobinState{}, // default - will be unused
	)
	if prioritizeQueryComponents {
		tree, err = NewTree(
			&roundRobinState{}, // select query component
			tqas,              // select tenant
			&roundRobinState{},// default - will be unused
		)
```

This is not essential, and more of an aesthetic change/idea but it may be a little more clear - since we don't use any algorithm at the lowest level of the tree we could set that to nil and have some sort of special treatment of these leaf nodes - nil algorithm, no child nodes allowed etc.

Reply:
I've rewritten these comments so they both describe what the node _is_, and what its algorithm _does_. Does that help?